include/rmm/device_uvector.hpp

```diff
@@ -0,0 +1,270 @@
+ * default initialize the elements. This initialization is often unnecessary, e.g., when the vector
+ * is created to hold some output from some operation.
+ *
+ * However, this restricts the element type `T` to only trivially copyable types. In short,
+ * trivially copyable types can be safely copied with `memcpy`. For more information, see
+ * https://en.cppreference.com/w/cpp/types/is_trivially_copyable.
+ *
+ * Another key difference over `thrust::device_vector` is that all operations that invoke
+ * allocation, kernels, or memcpys take a CUDA stream parameter to indicate on which stream the
+ * operation will be performed.
+ *
+ * @tparam T Trivially copyable element type
+ */
+template <typename T>
+class device_uvector {
+  static_assert(std::is_trivially_copyable<T>::value,
+                "device_uvector only supports types that are trivially copyable.");
+
+ public:
+  using value_type     = T;
+  using size_type      = std::size_t;
+  using reference      = value_type&;
+  using pointer        = value_type*;
+  using const_pointer  = value_type const*;
+  using iterator       = pointer;
+  using const_iterator = const_pointer;
+
+  device_uvector()                      = default;
+  ~device_uvector()                     = default;
+  device_uvector(device_uvector const&) = default;
+  device_uvector(device_uvector&&)      = default;
+  device_uvector& operator=(device_uvector const&) = default;
+  device_uvector& operator=(device_uvector&&) = default;
+
+  /**
+   * @brief Construct a new `device_uvector` with sufficient uninitialized storage for `size`
+   * elements.
+   *
+   * Elements are uninitialized. Reading an element before it is initialized results in undefined
+   * behavior.
+   *
+   * @param size The number of elements to allocate storage for
+   * @param stream The stream on which to perform the allocation
+   * @param mr The resource used to allocate the device storage
+   */
+  explicit device_uvector(std::size_t size,
+                          cudaStream_t stream                 = 0,
+                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    : _storage{size * sizeof(T), stream, mr}, _size{size} {}
+
+  /**
+   * @brief Construct a new `device_uvector` with the specified number of elements initialized to
+   * the specified value.
+   *
+   * @param size The number of elements
+   * @param v The value used to initialize all elements
+   * @param stream The stream on which to perform the allocation and initialization
+   * @param mr The resource used to allocate the device storage
+   */
+  device_uvector(std::size_t size,
+                 value_type const& v,
+                 cudaStream_t stream                 = 0,
+                 rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+    : device_uvector(size, stream, mr) {
+    if (std::is_integral<T>::value && (value_type{0} == v)) {
```

Comment:
This should work for float types too, so should this be `std::enable_if_t<std::is_fundamental<T>::value`?  This is what I used in `device_scalar`.

On the other hand, shouldn't `thrust::uninitialized_fill` be making this optimization anyway?

Reply:
> This should work for float types too, so should this be `std::enable_if_t<std::is_fundamental<T>::value`?

Good point, though I think it should be `is_arithmetic`. `is_fundamental` includes `nullptr_t`, which there is no guarantee that `nullptr` is "all zero bits". 

> On the other hand, shouldn't thrust::uninitialized_fill be making this optimization anyway?

Potentially. There would need to be some extra checks to see when the iterators passed to `uninitialized_fill` are raw data pointers and not some output iterator type.